src/main/java/games/strategy/util/memento/PropertyBagMemento.java

```diff
@@ -0,0 +1,84 @@
+package games.strategy.util.memento;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Objects;
+
+/**
+ * Implementation of {@link Memento} that represents the originator state as a collection of name-value pairs.
+ *
+ * <p>
+ * Instances of this class are immutable.
+ * </p>
+ */
+public final class PropertyBagMemento implements Memento {
```

Comment:
On another topic, if I understand `PropertyBag` correctly, we basically are abstracting any class we need to proxy as a set of key-value properties. This makes sense, the concept of a `PropertyBag` maps logically well to viewing a class as set of key-value properties. The only question I would have is how are nested types handled. Do they need their own memento, are they supported, or is support for them automatic? I actually suspect if I traced further and in more detail this down would be more obvious, and the framework usage of a `PropertyBagMemento` would be more obvious. Failing that, the issues I am running into is it is not quite clear how this class fits into the framework, and what its relationship would be with GameData objects. Is this class 1:1 to any other class, or is it a stand-alone helper  in the  framework?